staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -798,6 +867,12 @@ func (alloc *allocator) allocateOne(r deviceIndices, allocateSubRequest bool) (b
 			return false, nil
 		}
 		done, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, deviceIndex: r.deviceIndex + 1}, allocateSubRequest)
+		// If we hit the allocation size limit, just release the
+		// allocated device and backtrack.
+		if errors.Is(err, errAllocationResultMaxSizeExceeded) {
+			deallocate()
+			return false, err
+		}
```

Comment:
So basically the decision here is:
- if the error is potentially something that isn't fatal, revert to the previous state
- otherwise we know that allocation has failed and don't clean up

Previously all errors were fatal, so there was no need to call `deallocate` when returning an error.

Would the logic become simpler with a `defer deallocate()` directly after allocating a device? Optimizing for a fast final failure path is pointless, so it shouldn't matter for performance.

It gets rid of if checks which currently need lots of comments to justify their existence...

Reply:
Let's always call `deallocate()` before all returns, unconditionally. Then we don't need to explain why we do it, because it is "normal".

Reply:
> I'm a bit uncertain if I feel the copying the allocation on success is needed

Not at the moment, it just makes returns simpler. Long-term it could be extended to collect multiple results.

Reply:
So I'm also getting a slightly worse result with the simple patch than the current PR. Agree that it is not obvious why it is worse.

However, compared to master I'm seeing a slight improvement even with the simple patch.

I suggest we use just the simple patch. I think the requirement to call deallocate when allocation has succeeded and we didn't find a solution is reasonable.

Reply:
Agreed, please add it.